### PR TITLE
Add mini-blog about sliced visiting

### PR DESCRIPTION
**Pull request by jonmv**
@kkraune and @jobergum please review.

### CODE REVIEW

File: _posts/2021-11-12-sliced-visiting.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+<p class="image-credit">
+ Photo by <a href="https://unsplash.com/@anniespratt?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Annie Spratt</a> on
+ <a href="https://unsplash.com/s/photos/cooperation-nature?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
+ </p>
+
+Vespa is built to scale; documents are automatically spread across any number of content nodes, and queries and document
+operations pass concurrently through the container layer in front. The container `/document/v1` API is intended for reading
+and writing documents, but while it supports high-throughput [feeding](https://docs.vespa.ai/en/vespa-feed-client.html),
+it has only provided limited throughput for bulk reads, i.e., _visiting_, that is, until now!
+This short blog post shows how the new slicing feature lets
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:23]:
```suggestion
This blog post shows how the new slicing feature lets
```

File: _posts/2021-11-12-sliced-visiting.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+excerpt: The new slicing feature in /document/v1 splits visiting across independent HTTP requests, letting throughput scale with the number of container nodes or clients.
+---
+
+![Decorative image](/assets/2021-11-12-sliced-visiting/annie-spratt-gSQbip5HHuc-unsplash.jpg)
+<p class="image-credit">
+ Photo by <a href="https://unsplash.com/@anniespratt?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Annie Spratt</a> on
+ <a href="https://unsplash.com/s/photos/cooperation-nature?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
+ </p>
+
+Vespa is built to scale; documents are automatically spread across any number of content nodes, and queries and document
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:19]:
```suggestion
Vespa is built to scale; documents are distributed across any number of content nodes, and queries and document
```

File: _posts/2021-11-12-sliced-visiting.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+a single visitor is limited by the HTTP throughput of a single container node, which is typically much lower than the
+throughout from the content nodes to this container. We therefore added the option to specify a number of
+[slices](https://docsc.vespa.ai/en/reference/document-v1-api-reference.html#slices) and a
+[sliceId](https://docsc.vespa.ai/en/reference/document-v1-api-reference.html#sliceId) when using `/document/v1`.
+This splits the corpus in the specified number of independent slices, which lets multiple visitors work concurrently,
+but independently, to advance the same logical visit. Finally, since HTTP requests are one-to-one with visitors, we can
+distribute a single logical visit across any number of container nodes, and also clients. A sample script for dumping the
+entire corpus is found in the Vespa [documentation](https://docs.vespa.ai/en/document-v1-api-guide.html#data-dump).
+
+## Experimental Throughput
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:47]:
```suggestion
## Experimental Throughput Benchmarking
```

File: _posts/2021-11-12-sliced-visiting.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+numbers were 17MB/s vs 399MB/s. In both cases, CPU utilisation increased by around 6 vCPU. 
+
+The reason for the worse performance for the coast-to-coast setup, prior to slicing, is that only parts of the corpus
+are contained in each HTTP request, and the next HTTP request cannot be sent before the previous response has been processed,
+and the [continuation token](https://docs.vespa.ai/en/reference/document-v1-api-reference.html#continuation) extracted. 
+As part of the slicing work, we also added the option to
+[stream](https://docs.vespa.ai/en/reference/document-v1-api-reference.html#stream) the HTTP responses, allowing much larger
+payloads and lower TTFB. Simply turning on this increased throughput to 27MB/s for the coast-to-coast setup, without slicing;
+it also increased throughput by 10%-20% when slicing was used, and was enabled for these measurements. 
+
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:66]:
```suggestion
# Summary
```

File: _posts/2021-11-12-sliced-visiting.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ <a href="https://unsplash.com/s/photos/cooperation-nature?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
+ </p>
+
+Vespa is built to scale; documents are automatically spread across any number of content nodes, and queries and document
+operations pass concurrently through the container layer in front. The container `/document/v1` API is intended for reading
+and writing documents, but while it supports high-throughput [feeding](https://docs.vespa.ai/en/vespa-feed-client.html),
+it has only provided limited throughput for bulk reads, i.e., _visiting_, that is, until now!
+This short blog post shows how the new slicing feature lets
+[visiting throughput](https://docs.vespa.ai/en/document-v1-api-guide.html#visiting-throughput) scale with the number of
+containers and the number of clients used for visiting—whicheveris the bottleneck.
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:25]:
```suggestion
containers and the number of clients used for visiting, whichever is the bottleneck.
```

File: _posts/2021-11-12-sliced-visiting.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+# Visiting Vespa
+[Visiting](https://docs.vespa.ai/en/content/visiting.html) iterates through all documents stored on the Vespa content nodes,
+optionally filtered by a [selection](https://docs.vespa.ai/en/reference/document-select-language.html), and sends these to
```
Comment (jobergum) [inline _posts/2021-11-12-sliced-visiting.md:29]: Hard to read this sentence, consider breaking it up.

**Review by jobergum — CHANGES_REQUESTED** [review]
Great! I would change the title to something more catchy

Other
- rewrite MB/s to network units e.g Gbps 
- avoid us/we (Yes, I happen to do this as well)
- Some long sentences
-

**Comment by kkraune** [comment]
great post! I think it would be much easier to understand if it had an illustration with multiple clients, container and content nodes, so you could see how the clients could use different container nodes to visit its subspace. maybe a link to https://docs.vespa.ai/en/content/buckets.html for details

also, the stream feature should be more prominent. it solves a different problem than slicing does, and should have its own section.

then it should all come together at the end, how the slicing adds throughput, and stream makes latency less of an issue, to achieve high read rates using multiple clients, over long distances (e.g. to support easy multizone setup, easy to transport the data using this)

**Comment by jonmv** [comment]
In my mind, the stream part is a minor thing. Using more slices makes it less impactful, as you'll always have data to process at both ends, and then it's sort of just a prerequisite for slicing. Playing it down was on purpose, to keep the post as short as possible. I also hid the buckets on purpose, as slicing isn't necessarily tied to those—it just happens to be, but that's a detail. Knowing exactly how slicing works isn't necessary to appreciate it, in my opinion, and I prefer keeping things simpler, with a smaller scope, when possible. 
I was considering a figure, though. Will see if I can make one I'm happy with. 
Let's keep the discussion going if you think I'm wrong about the above points.

**Comment by kkraune** [comment]
you are too modest :-) both slicing and streaming are means to achieve higher throughput, where streaming is the easiest once for the user, but less impactful, unless in the high-latency case. I will not insist, but I thought streaming is important enough for a more prominent position. lets have @jobergum be the tiebreaker ;-)

**Comment by jonmv** [comment]
On second thought, a figure doesn't make that much sense either without the buckets being disjoint across content nodes, etc., either, as it'll just be a few clients hitting different containers.

**Comment by jonmv** [comment]
Hmm, yes, it is a point that streamed responses are beneficial also when you don't want to use parallel clients, and slicing. So I tried to put it in a separate section at first, but wasn't happy with that, because I found it took too much (cognitive) space, since it also entails a description of continuation tokens, which again are needed because visits are stateless between HTTP calls, and so on. But I could make another attempt at this ...

**Comment by jonmv** [comment]
PTAL.

**Comment by kkraune** [comment]
Great! I like the illustration! and the text flow is much better. maybe just rename to today's date and ship it :-)

**Review by kkraune — APPROVED** [review]

**Pull request merged**